pkg/soc/soc.go

```diff
@@ -0,0 +1,264 @@
+func (s *Chunk) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
+}
+
+// FromChunk recreates an Chunk from swarm.Chunk data.
+func FromChunk(ch swarm.Chunk) (*Chunk, error) {
+	chunkData := ch.Data()
+	if len(chunkData) < minChunkSize {
+		return nil, errors.New("less than minimum length")
+	}
+
+	// add all the data fields
+	sch := &Chunk{}
+	cursor := 0
+
+	sch.id = chunkData[cursor : cursor+IdSize]
+	cursor += IdSize
+
+	sch.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
+
+	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
+	span := binary.LittleEndian.Uint64(spanBytes)
+	sch.span = int64(span)
+	cursor += swarm.SpanSize
+
+	sch.payload = chunkData[cursor:]
+
+	bmtPool := bmtlegacy.NewTreePool(swarm.NewHasher, swarm.Branches, bmtlegacy.PoolSize)
+	bmtHasher := bmtlegacy.New(bmtPool)
+
+	// calculate the bmt hash of the sch payload
+	err := bmtHasher.SetSpan(int64(span))
+	if err != nil {
+		return nil, err
+	}
+	_, err = bmtHasher.Write(sch.payload)
+	if err != nil {
+		return nil, err
+	}
+	payloadSum := bmtHasher.Sum(nil)
+
+	sha3Hasher := swarm.NewHasher()
+	_, err = sha3Hasher.Write(payloadSum)
+	if err != nil {
+		return nil, err
+	}
+	_, err = sha3Hasher.Write(sch.id)
+	if err != nil {
+		return nil, err
+	}
+	toSignBytes := sha3Hasher.Sum(nil)
+
+	// recover owner information
+	recoveredPublicKey, err := crypto.Recover(sch.signature, toSignBytes)
+	if err != nil {
+		return nil, err
+	}
+	recoveredEthereumAddress, err := crypto.NewEthereumAddress(*recoveredPublicKey)
+	if err != nil {
+		return nil, err
+	}
+	owner, err := NewOwner(recoveredEthereumAddress)
+	if err != nil {
+		return nil, err
+	}
+	sch.owner = owner
+
+	return sch, nil
+}
+
+// CreateChunk creates a new chunk with signed payload ready for submission to the swarm network
+// from the given update data.
+//
+// This method will fail if no signer has been defined.
+func (s *Chunk) CreateChunk() (swarm.Chunk, error) {
+	if s.signer == nil {
+		return nil, errors.New("signer missing")
+	}
+	bmtPool := bmtlegacy.NewTreePool(swarm.NewHasher, swarm.Branches, bmtlegacy.PoolSize)
+	bmtHasher := bmtlegacy.New(bmtPool)
+
+	// calculate the bmt hash of the update payload
+	err := bmtHasher.SetSpan(s.span)
+	if err != nil {
+		return nil, err
+	}
+	_, err = bmtHasher.Write(s.payload)
+	if err != nil {
+		return nil, err
+	}
+	payloadSum := bmtHasher.Sum(nil)
+	sha3Hasher := swarm.NewHasher()
+	_, err = sha3Hasher.Write(payloadSum)
+	if err != nil {
+		return nil, err
+	}
+	_, err = sha3Hasher.Write(s.id)
+	if err != nil {
+		return nil, err
+	}
+	toSignBytes := sha3Hasher.Sum(nil)
+
+	// sign the update
+	signature, err := s.signer.Sign(toSignBytes)
```

Comment:
i. dont think you need to hash them together, you can just use 

     toSignBytes. = bmtHash | ID

Reply:
Reducing to a hash before signing is a fairly common concept, though?

Reply:
I think it's a fairly common construct to sign digests rather than arbitrary length bytes.

In any case, the signer implementation seems to be acting up when trying to sign this concatenation. 

May I suggest that you adjust the spec to specify signature is digest instead?

@zelig

Reply:
yes but the `signer.Sign` does that anyway no?

Reply:
doesnt seem to.